orc/src/main/java/com/netflix/iceberg/orc/ORC.java

```diff
@@ -0,0 +1,154 @@
+/*
+ * Copyright 2018 Hortonworks
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.netflix.iceberg.orc;
+
+import com.google.common.base.Preconditions;
+import com.netflix.iceberg.PartitionSpec;
+import com.netflix.iceberg.Schema;
+import com.netflix.iceberg.io.InputFile;
+import com.netflix.iceberg.io.OutputFile;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.fs.Path;
+import org.apache.orc.OrcFile;
+import org.apache.orc.Reader;
+import org.apache.orc.TypeDescription;
+
+import java.io.IOException;
+import java.nio.charset.StandardCharsets;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+
+public class ORC {
+  private ORC() {
+  }
+
+  public static WriteBuilder write(OutputFile file) {
+    return new WriteBuilder(file);
+  }
+
+  public static class WriteBuilder {
+    private final OutputFile file;
+    private Schema schema = null;
+    private PartitionSpec spec = null;
+    private Configuration conf = null;
+    private final Properties tableProperties = new Properties();
+    private Map<String, byte[]> metadata = new HashMap<>();
+
+    private WriteBuilder(OutputFile file) {
+      this.file = file;
+    }
+
+    public WriteBuilder partitionSpec(PartitionSpec spec) {
+      this.spec = spec;
+      return this;
+    }
+
+    public WriteBuilder metadata(String property, String value) {
+      metadata.put(property, value.getBytes(StandardCharsets.UTF_8));
+      return this;
+    }
+
+    public WriteBuilder tableProperties(Properties properties) {
+      tableProperties.putAll(properties);
+      return this;
+    }
+
+    public WriteBuilder schema(Schema schema) {
+      this.schema = schema;
+      return this;
+    }
+
+    public WriteBuilder conf(Configuration conf) {
```

Comment:
I'm trying to avoid Hadoop classes in APIs like this (even if they aren't in the iceberg-api module). In Parquet, we added `config` that adds properties to the internal `Configuration`. That configuration is passed in as part of a `HadoopOutputFile` instance. If `HadoopOutputFile` isn't used, then the writer should fall back to `new Configuration()` if it is required.

Reply:
Creating new Configuration object is very expensive. You really want to avoid doing at all costs. And it limits what the configuration to the default.  ORC's API needs a Hadoop Configuration. So if I don't pass one down, I'll need to create it and deprive the user of the ability to pass down a context specific Configuration.

Reply:
I think the right way to pass that configuration is through a `HadoopOutputFile`. If there is a `Configuration` to pass, then the file passed in will be a Hadoop one and you can use it. If it isn't, then you'd have to create a configuration anyway in order to use ORC because it requires one. I guess a better way to describe it is not avoiding Hadoop classes, but keeping them in Hadoop-specific areas (not in API) so we can possibly remove the classes later.

For this, I think just moving to a `config(String, String)` method is a good idea, and using those configs to update a `Configuration` from the `HadoopOutputFile` instance.

Reply:
I can do that, but I don't see why making a strong dependence on HadoopOutputFile and HadoopInputFile is ok, but Configuration is bad. In either case, you aren't going to be able to compile or run the Iceberg code without hadoop-common on the class path.

Reply:
It just partitions Hadoop classes in their own package so that we don't leak them in the main parts of the API. The eventual goal is to support tables without pulling in Hadoop.